### PR TITLE
Introducing dynamic networks

### PR DESCRIPTION
**Pull request by LucasFuentesValenzuela**
Aim is to introduce dynamic networks and implicit diff in those problems.

### CODE REVIEW

File: src/dyn_model.jl [hunk 1]
```diff
@@ -0,0 +1,307 @@
+# Code for building and running a power management problem
+
+
+# ===
+# POWER MANAGEMENT PROBLEM
+# ===
+
+TAU = 1e-5
```
Comment (degleris1) [inline src/dyn_model.jl:8]: Can we set `TAU = 0.0` to start?

File: src/dyn_model.jl [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+
+TODO: check that we're not messing up the ordering!
+"""
+function PowerManagementProblem(fq, fl, d, pmax, gmax, A, ds; τ=TAU)
+    n, m = size(A)
+    g = Variable(n)
+    p = Variable(m)
+
+    # add variable-specific constraints
+    add_constraints(p, [
```
Comment (degleris1) [inline src/dyn_model.jl:41]:
```suggestion
    add_constraints!(p, [
```

File: src/dyn_model.jl [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+        -p <= pmax, 
+        p <= pmax, 
+    ])
+    add_constraints!(g, [
```
Comment (degleris1) [inline src/dyn_model.jl:45]: I think `add_constraints!` only works for `Convex.Problem` objects. We might need to use `add_constraint!` and add constraints one at a time for `Convex.Variable` objects. Need to double check.

File: src/dyn_model.jl [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+    )
+
+    # build problem
+    problem = minimize(
+        (1/2)*quadform(g, Diagonal(fq))
+        + fl'g
+        + (τ/2)*sumsquares(p)
+    )
+    # add network constraint
+    add_constraint!(problem, 0 == A*p - g + d + ds)
```
Comment (degleris1) [inline src/dyn_model.jl:58]: See my most recent two commits---I think we should support this `B` matrix (mapping generators to nodes) here. It has made the synthetic analysis much more interesting, I think.

```suggestion
    add_constraint!(problem, 0 == A*p - B*g + d + ds)
```

File: src/dyn_model.jl [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+
+The arguments to DynPowerManagementProblem are the same as for PowerManagementProblem
+except they are arrays, with dimension equal to the time horizon (i.e. number of timesteps) T. 
+
+P is the maximum charge/discharge power. 
+C is the maximum SOC of the batteries.
+
+TODO: parameterize the management of initial and terminal constraints
+"""
+function DynPowerManagementProblem(
```
Comment (degleris1) [inline src/dyn_model.jl:80]: Being picky here. In Julia, the convention has been to (almost always) use the full name. I don't think `DynamicPowerManagementProblem` is too much worse. Feel free to ignore this suggestion though.

```suggestion
function DynamicPowerManagementProblem(
```

File: src/dyn_model.jl [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+    fq, fl, d, pmax, gmax, A, P, C; τ=TAU
+)
+    T = length(fq)
+    n, _ = size(A)
+
+    # Define a storage variable for n nodes over T timesteps
+    # TODO: for now the edge constraints are neglected
+    s = [Variable(n) for t in 1:T]
+    add_constraints!(s,[
```
Comment (degleris1) [inline src/dyn_model.jl:89]: You'll have to do `append!([0 <= s[t] for t in 1:T], [s[t] <= C for t in 1:T])`

File: src/dyn_model.jl [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+        0 <= s <= C
+        # s[:, 1] == 0, 
+        # s[:, T+1] == 0
+    ])
+    add_constraints!(s, [
+        s[t+1] - s[t] <= P for t in 1:T
```
Comment (degleris1) [inline src/dyn_model.jl:95]: This will throw an error for `t=T`

**Comment by LucasFuentesValenzuela** [comment]
@degleris1 I changed a bit the structure. Can you quickly give another look? 

I checked that it runs. The problem is not yet feasible apparently -- so some debugging to do. But I am kind of happy with the structure so let me know.

**Comment by LucasFuentesValenzuela** [comment]
@degleris1 I have tried to have the dynamic problem work on a very simple setup (i.e. small network, one timestep). This is in notebook `dyn_mef_synthetic`. Current version of the code actually has elements related to `s` (i.e. the storage value) commented out and for one timestep I am expecting to get **exactly** the same result as the static network. Can I ask you to quickly check the code/play a bit with the notebook to see if you can understand where it fails? Normally you would have to just play with section 3.1. 

I have not been able to diagnose it. It looks like it's something simple I am missing. 

Thanks

File: src/model.jl [hunk 8]
```diff
@@ -53,12 +54,12 @@ function PowerManagementProblem(fq, fl, d, pmax, gmax, A, B; τ=TAU)
...
         -g <= 0, 
         g <= gmax,
-        0 == A*p - B*g + d,
+        0 == A*p - B*g + d + ds,
     ])
 
     params = (fq=fq, fl=fl, d=d, pmax=pmax, gmax=gmax, A=A, B=B, τ=τ)
 
-    return PowerManagementProblem(problem, p, g, params)
+    return PowerManagementProblem(problem, p, g, nothing, params)
```
Comment (degleris1) [inline src/model.jl:62]:
```suggestion
    return PowerManagementProblem(problem, p, g, zeros(n), params)
```

File: src/model.jl [hunk 9]
```diff
@@ -53,12 +54,12 @@ function PowerManagementProblem(fq, fl, d, pmax, gmax, A, B; τ=TAU)
...
 end
 
 PowerManagementProblem(net::PowerNetwork, d) =
```
Comment (degleris1) [inline src/model.jl:65]: Does this need to be changed too?

Comment (LucasFuentesValenzuela) [reply src/model.jl:65]: You're right!

Comment (LucasFuentesValenzuela) [reply src/model.jl:65]: well, I guess it's not necessary per se. I just thought that having only one `struct` for individual subproblems would be nice, and therefore I repurposed the `PowerManagementProblem` one to allow for this flexibility. If you don't like it, I can find another way

File: src/CarbonNetworks.jl [hunk 10]
```diff
@@ -1,6 +1,7 @@
 module CarbonNetworks
 
 # Module Imports
+using Zygote: length
```
[the remark this answers is not in the document]
Comment (LucasFuentesValenzuela) [reply src/CarbonNetworks.jl:4]: Not sure actually.

File: src/dyn_model.jl [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+    DynamicPowerManagementProblem(fq, fl, d, pmax, gmax, A, B, P, C; τ=TAU)
+
+The arguments to DynamicPowerManagementProblem are the same as for PowerManagementProblem
+except `fq, fl, d, pmax, gmax` are arrays of dimension `T`, where `T` is
+the time horizon. 
+
+Additional arguments are:
+- `P`: the maximum charge/discharge power. 
+- `C`: the maximum SOC of the batteries.
+"""
```
Comment (LucasFuentesValenzuela) [inline src/dyn_model.jl:58]: Standardize doc

File: src/dyn_model.jl [hunk 12]
```diff
@@ -0,0 +1,383 @@
...
+function DynamicPowerManagementProblem(
+    fq, fl, d, pmax, gmax, A, B, P, C; τ=TAU
+)
+    T = length(fq)
+    n, _ = size(A)
+
+    # Define a storage variable for n nodes over T timesteps
+    # TODO: for now the edge constraints are neglected
```
Comment (LucasFuentesValenzuela) [inline src/dyn_model.jl:66]: remove and clarify

File: src/dyn_model.jl [hunk 13]
```diff
@@ -0,0 +1,383 @@
...
+"""
+function flatten_variables_dyn(P::PowerManagementProblem)
+
+    T = length(P.g)
+    n, m = size(P.params.A)
+    _, l = size(P.params.B)
+
+    x_static = Array{Float64}(undef, 0)
+    x_storage = Array{Float64}(undef, 0)
+
```
Comment (LucasFuentesValenzuela) [inline src/dyn_model.jl:292]: @degleris1 I saw you changed this in the above to `Float64[]`. Is that the standard way to write it?

**Review by LucasFuentesValenzuela — COMMENTED** [review]
Looks good, just minor stuff

**Pull request merged**